Code Adventures Simulation 3D

jamesm2w opened the pull request:
Code Adventures simulation upgraded to the third dimensions. 

There still might need to be some parameter tuning in the future, but it works for a fairly basic simulation right now. 

Performance is definitely an issue when running this one.

mwasacz reviewed (commented):
Looks great!
Personally I'd replace the `#define`s with constants/functions, the compiler should be able to inline them anyway. But I'm not a C++ expert, so let's wait for Scott's opinion.

jamesm2w commented:
> Looks great! Personally I'd replace the `#define`s with constants/functions, the compiler should be able to inline them anyway. But I'm not a C++ expert, so let's wait for Scott's opinion.

Yeah, I don't think there particularly any difference, just mainly for ease of use to be able to change the parameters in one place. I do think that it can't really optimise the `add_variable` calls and that might be slowing down performance compared to #define-ing the values as in the 2D simulation.

Scott943 commented:
> Looks great! Personally I'd replace the `#define`s with constants/functions, the compiler should be able to inline them anyway. But I'm not a C++ expert, so let's wait for Scott's opinion.

This is a good point. `#define` macros are great for conditional compilation, but they are not typed. Instead they literally do a search and replace on the source code, which leads to less robust code. I would personally prefer constants to be defined with the `constexpr` qualifer, which which means they can still be resolved at compile time _and_ they are typed. Additionally, I'm not too sure about having a macro that executes functions - not too sure why this can't just use a regular function call.

Scott943 commented on src/examples/3d_fluid_sim/fluid_sim.cpp:
```diff
@@ -0,0 +1,198 @@
+    wash::add_variable("pressureMultiplier", 288.0); \
+    wash::add_variable("nearPressureMultiplier", 2.25); \
+    wash::add_variable("viscosityStrength", 0.08);
+#define numParticles 42875
```
I would declare these as e.g. `constexpr uint64 numParticles = 42875;`

jamesm2w replied:
This was a fairly non-trivial transformation involving updating Vec constructors to be constexpr, and adding a new const Vec multiplication operator override (probably have to eventually implement more of those) but done. This should probably help the performance as well.

Scott943 commented on src/examples/3d_fluid_sim/fluid_sim.cpp:
```diff
@@ -0,0 +1,198 @@
+}
+
+void pressure(wash::Particle& particle, const std::vector<wash::Particle>& neighbours) {
+    auto pos = particle.get_force_vector("predictedCoordinates");
```
For these ones that aren't changing over the course of the function, you can mark them as `const auto`

jamesm2w replied:
See previous comment, transform to constexpr globals made.

Scott943 commented on src/examples/3d_fluid_sim/fluid_sim.hpp:
```diff
@@ -0,0 +1,86 @@
+#define DIM 3
+#endif
+
+#include <random>
```
Minor: I personally put all `include`s before `define`s

jamesm2w replied:
It was done this way to avoid getting the default initialisation of DIM to be 2 from the vector (I think) header file

Scott943 commented on src/examples/3d_fluid_sim/fluid_sim.hpp:
```diff
@@ -0,0 +1,86 @@
+#endif
+
+#if DIM == 2
+#define DIM 3
```
I'd prefer to error here if the wrong DIM is specified for this simulation

jamesm2w replied:
Done. Uses an error directive now.

Scott943 commented on src/examples/3d_fluid_sim/fluid_sim.hpp:
```diff
@@ -0,0 +1,86 @@
+    auto mag = pos.magnitude();
+
+    if (mag == 0.0) {
+        return wash::Vec3D{0.0, 0.0, 0.0};
```
Should be able to write `wash::Vec3D{}` as default initialiser is all zeroes

jamesm2w replied:
Changed.

jamesm2w commented:
> > Looks great! Personally I'd replace the `#define`s with constants/functions, the compiler should be able to inline them anyway. But I'm not a C++ expert, so let's wait for Scott's opinion.
> 
> This is a good point. `#define` macros are great for conditional compilation, but they are not typed so aren't great for variable declarations. Instead they literally do a search and replace on the source code, which leads to less robust code. I would personally prefer constants to be defined with the `constexpr` qualifer, which which means they can still be resolved at compile time _and_ they are typed. If the declaration is a object that doesn't work with `constexpr`, `const` is normally good enough too.
> 
> Additionally, I'm not too sure about having a macro that executes functions - not too sure why `DECLARE_VARIABLES` can't just use a regular function call.
> 
> Great job overall @jamesm2w

I basically wanted the find-and-replace semantics tbf, but if compile time replacement works with `constexpr` I've updated it to use that. Additionally, updated the constructors of vec to be constexpr, let me know if that's been done wrong or if there's a better way to do it. 

The `DECLARE_VARIABLES` macro was just to define all the parameters in one place rather than have to look in two places to quickly change simulation parameters, either way I've updated those to be constant globals now anyway.
